20231214-listUsers-api.md

```diff
@@ -0,0 +1,841 @@
+
+* **Userset** - A userset refers to a set of zero or more concrete objects that are the subjects that a relationship expands to. See our Concepts page on [Users](https://openfga.dev/docs/concepts#what-is-a-user) for more info.
+
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+Developers using OpenFGA want to be able to do a reverse query and lookup all of the users that have a particular relationship with an object. 
+
+- What use cases does it support?
+
+**UIs** - display the users that a resource has been shared with. Think of the "Share" dialog in Google Docs, for example.
+
+
+- What is the expected outcome?
+
+Two new core APIs are added to the [OpenFGA API](https://github.com/openfga/api) and implementations of these new APIs in the [OpenFGA server](https://github.com/openfga/openfga).
+
+# What it is
+[what-it-is]: #what-it-is
+
+Given an `object`, `relation`, and one or more user/subject provided filters, return the concrete/terminal set of user objects matching at least one of the user filters that have that relationship with the object.
+
+More formally, given the following model and relationship tuples:
+
+```
+model
+  schema 1.1
+
+type user
+type cat
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [cat, user, group#member]
+```
+
+| object     | relation | user             | condition_name | context |
+|------------|----------|------------------|----------------|---------|
+| document:1 | viewer   | user:anne        | null           | null    |
+| document:1 | viewer   | group:eng#member | null           | null    |
+| group:eng  | member   | group:fga#member | null           | null    |
+| group:fga  | member   | user:jon         | null           | null    |
+
+ListUsers and StreamedListUsers would behave as follows:
+
+> ℹ️ For the sake of brevity the examples below focus on the ListUsers unary RPC variant. The difference with the StreamedListUsers API is the streaming semantics of the response. Instead of returning an array, the streaming variant returns each result one at a time as they are "discovered".
+
+**Example 1**
+`user` object types are both directly and indirectly related to `document#viewer` relationships. The indirect relationships come through the direct relationships involving `group#member` (e.g. users can directly view a document or if the user is part of a group that can view the document)
+```
+ListUsers({
+    object: "document:1",
+    relation: "viewer",
+    user_filter: [{type: "user"}]
```

Comment:
why is this an array and not just one string like ListObjects? is there a hidden use case here?